packages/compiler-core/src/transforms/vOn.ts

```diff
@@ -35,7 +35,13 @@ export const transformOn: DirectiveTransform = (
   augmentor
 ) => {
   const { loc, modifiers, arg } = dir as VOnDirectiveNode
-  if (!dir.exp && !modifiers.length) {
+  let exp: ExpressionNode | undefined = dir.exp as
+    | SimpleExpressionNode
+    | undefined
+  if (exp && !exp.content.trim()) {
+    exp = undefined
+  }
+  if (!exp && !modifiers.length) {
```

Comment:
This check **should** be performed before setting empty exp to `undefined` because empty `v-on` with modifiers are allowed, e.g. `@click.prevent`